rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2600 @@
+
+**blocks:**
+
+- have greater confidence that the data passed to them is correct and useable
+- are able to be more specific about their requirements and more easily express the function of the data
+  - This also facilitates more powerful composition as there's greater confidence about other parties' functionalities.
+
+**Data Producers** (whether that be end-users, EA developers, or businesses):
+
+- have a toolkit to better describe the data they're producing and to help them be explicit about semantic meaning
+- benefit from a community-driven description of knowledge, potentially removing barriers of communication between domains
+
+# Guide-Level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+## Goals
+
+Given the motivation above and some other implications of the current design, this RFC attempts to:
+
+1.  make descriptions of data **discoverable**, **reusable**, and **composable**
+1.  remove the few instances where the specification currently misuses (modifies in a breaking way) JSON schema
+
+As well as addressing these existing shortcomings, this RFC seeks to:
+
+<!-- markdownlint-disable  MD029 -->
+
+3.  define a constrained type system that is able to describe any JSON structure
+4.  refine the current way of defining links between entities, encapsulating the approach within the type system
+
+<!-- markdownlint-enable  MD029 -->
+
+## Types
+
+Underpinning this proposal is the definition of four classes of "types". One of these classes is an extension of the current definition of an _Entity Type_ while the other three are new classifications that are designed to improve the usability and expressiveness of _Entity Types_.
+
+### Data Types
+
+A **Data Type** describes a space of possible valid _values_. For instance, the _string_ value `"foo"` might be an instance of a `Text` Data Type, or a _float_ value `0.6` might be an instance of a `Number` Data Type ("instance of" means that the value satisfies the constraints expressed on the Data Type, or alternatively you can view it as saying the value is within the value space of the Data Type).
+
+> 💬 The underlying value's representation will depend on the implementation environment, different languages and technologies may store these differently.
+
+A Data Type is composed of the following:
+
+- a **required** `$id`, which is a globally unique identifier, where the Data Type's definition can be accessed (in most cases this will be a URL)
+- a **required** `name`, which should generally be a non-pluralized short description of the Data Type (for example: "Number" not "Numbers")
+- **optionally** a `description` to further explain the semantic meaning of the Data Type
+- a **required** definition of its possible value
+
+#### Primitive Top-Level Data Types
+
+We pre-define a set of primitive top-level Data Types.
+These can be considered the least restrictive value spaces:
+
+- `Text`
+- `Number`
```

Comment:
Should we include `Integer` here? Or save that for user-created ones?

Reply:
My thinking is `Integer` is a more constrained version of `Number`, therefore should not be considered as one of our "primitive" data types. So yes, we should probably introduce it as part of the "user-created" data types.

Reply:
I'm inclined to agree, my only worry is that if user-generated types does take a while to be merged, then users can't actually specify an integer type until then.